sklearn/svm/src/liblinear/liblinear_helper.c

```diff
@@ -71,48 +63,49 @@ static struct feature_node **dense_to_sparse(double *x, npy_intp *dims,
  * Convert scipy.sparse.csr to libsvm's sparse data structure
  */
 static struct feature_node **csr_to_sparse(double *values, int *indices,
-        int *indptr, int n_samples, int n_features, double bias)
+        int *indptr, int n_samples, int n_features, int n_nonzero, double bias)
 {
-    struct feature_node **sparse, *temp;
+    struct feature_node **sparse;
     int i, j=0, k=0, n;
+    struct feature_node *T;
     int have_bias = (bias > 0);
 
-    sparse = malloc (n_samples * sizeof(struct feature_node *));
+    sparse = malloc (n_samples * sizeof(struct feature_node *) +
+                     n_nonzero * sizeof(struct feature_node) +
+                     ((have_bias+1) * n_samples) * sizeof(struct feature_node));
     if (sparse == NULL)
         return NULL;
 
+    T = (struct feature_node *)(sparse + n_samples);
```

Comment:
Does this provide much benefit? I fear it reduces readability.

Reply:
The significant performance improvement comes from moving the calls to malloc out of the loop. So yeah, we can have a separate malloc call here and avoid the cast. I just pushed a change to that effect.

Eventually I want to add a fast path for CSC sparse matrices, and it turns out that consolidating the malloc calls is especially important for converting a CSC matrix to liblinear format quickly. So I figured I should get that out of the way first before adding a csc_to_sparse function.